File: sdk-python/dephealth/dependency.py
"""Core abstractions: Dependency, Endpoint, CheckConfig, DependencyType."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import StrEnum


class DependencyType(StrEnum):
    """Dependency type."""

    HTTP = "http"
    GRPC = "grpc"
    TCP = "tcp"
    POSTGRES = "postgres"
    MYSQL = "mysql"
    REDIS = "redis"
    AMQP = "amqp"
    KAFKA = "kafka"
    LDAP = "ldap"


# Default values from the specification.
DEFAULT_CHECK_INTERVAL: float = 15.0
DEFAULT_TIMEOUT: float = 5.0
DEFAULT_INITIAL_DELAY: float = 5.0
DEFAULT_FAILURE_THRESHOLD: int = 1
DEFAULT_SUCCESS_THRESHOLD: int = 1

# Validation boundaries.
MIN_CHECK_INTERVAL: float = 1.0
MAX_CHECK_INTERVAL: float = 300.0
MIN_TIMEOUT: float = 1.0
MAX_TIMEOUT: float = 60.0
MIN_INITIAL_DELAY: float = 0.0
MAX_INITIAL_DELAY: float = 300.0
MIN_THRESHOLD: int = 1
MAX_THRESHOLD: int = 100

_NAME_PATTERN = re.compile(r"^[a-zA-Z][a-zA-Z0-9_-]{0,62}$")

LABEL_NAME_PATTERN = re.compile(r"^[a-zA-Z_][a-zA-Z0-9_]*$")

RESERVED_LABELS: frozenset[str] = frozenset(
    {"name", "group", "dependency", "type", "host", "port", "critical"}
)


def validate_label_name(label: str) -> None:
    """Validate a custom label name."""
    if not LABEL_NAME_PATTERN.match(label):
        msg = f"invalid label name {label!r}: must match [a-zA-Z_][a-zA-Z0-9_]*"
        raise ValueError(msg)
    if label in RESERVED_LABELS:
        msg = f"label name {label!r} is reserved"
        raise ValueError(msg)


def validate_labels(labels: dict[str, str]) -> None:
    """Validate all custom labels."""
    for key in labels:
        validate_label_name(key)


@dataclass(frozen=True)
class CheckConfig:
    """Dependency check configuration."""

    interval: float = DEFAULT_CHECK_INTERVAL
    timeout: float = DEFAULT_TIMEOUT
    initial_delay: float = DEFAULT_INITIAL_DELAY
    failure_threshold: int = DEFAULT_FAILURE_THRESHOLD
    success_threshold: int = DEFAULT_SUCCESS_THRESHOLD

    def validate(self) -> None:
        """Validate the configuration values."""
        if not MIN_CHECK_INTERVAL <= self.interval <= MAX_CHECK_INTERVAL:
            msg = f"interval must be between {MIN_CHECK_INTERVAL} and {MAX_CHECK_INTERVAL}"
            raise ValueError(msg)
        if not MIN_TIMEOUT <= self.timeout <= MAX_TIMEOUT:
            msg = f"timeout must be between {MIN_TIMEOUT} and {MAX_TIMEOUT}"
            raise ValueError(msg)
        if not MIN_INITIAL_DELAY <= self.initial_delay <= MAX_INITIAL_DELAY:
            msg = f"initial_delay must be between {MIN_INITIAL_DELAY} and {MAX_INITIAL_DELAY}"
            raise ValueError(msg)
        if not MIN_THRESHOLD <= self.failure_threshold <= MAX_THRESHOLD:
            msg = f"failure_threshold must be between {MIN_THRESHOLD} and {MAX_THRESHOLD}"
            raise ValueError(msg)
        if not MIN_THRESHOLD <= self.success_threshold <= MAX_THRESHOLD:
            msg = f"success_threshold must be between {MIN_THRESHOLD} and {MAX_THRESHOLD}"
            raise ValueError(msg)


def default_check_config() -> CheckConfig:
    """Return a configuration with default values."""
    return CheckConfig()


@dataclass(frozen=True)
class Endpoint:
    """Dependency endpoint address."""

    host: str
    port: str
    labels: dict[str, str] = field(default_factory=dict)


def bool_to_yes_no(value: bool) -> str:
    """Convert a bool to 'yes'/'no'."""
    return "yes" if value else "no"


@dataclass(frozen=True)
class Dependency:
    """Dependency descriptor."""

    name: str
    type: DependencyType
    critical: bool
    endpoints: list[Endpoint] = field(default_factory=list)
    config: CheckConfig = field(default_factory=default_check_config)

    def validate(self) -> None:
        """Validate the dependency descriptor."""
        validate_name(self.name)
        if not self.endpoints:
            msg = "at least one endpoint required"
            raise ValueError(msg)
        for ep in self.endpoints:
            validate_labels(ep.labels)
        self.config.validate()


def validate_name(name: str) -> None:
    """Validate a dependency name."""
    if not _NAME_PATTERN.match(name):
        msg = f"invalid dependency name {name!r}: must match [a-zA-Z][a-zA-Z0-9_-]{{{{0,62}}}}"
        raise ValueError(msg)

File: sdk-go/dephealth/dependency.go
// Package dephealth provides SDK for monitoring dependencies of microservices.
// Each dependency is checked periodically and exposes Prometheus metrics
// (app_dependency_health gauge and app_dependency_latency_seconds histogram).
package dephealth

import (
	"fmt"
	"regexp"
	"time"
)

// DependencyType represents the type of a dependency connection.
type DependencyType string

const (
	TypeHTTP     DependencyType = "http"
	TypeGRPC     DependencyType = "grpc"
	TypeTCP      DependencyType = "tcp"
	TypePostgres DependencyType = "postgres"
	TypeMySQL    DependencyType = "mysql"
	TypeRedis    DependencyType = "redis"
	TypeAMQP     DependencyType = "amqp"
	TypeKafka    DependencyType = "kafka"
)

// ValidTypes contains all valid dependency types.
var ValidTypes = map[DependencyType]bool{
	TypeHTTP:     true,
	TypeGRPC:     true,
	TypeTCP:      true,
	TypePostgres: true,
	TypeMySQL:    true,
	TypeRedis:    true,
	TypeAMQP:     true,
	TypeKafka:    true,
}

// Default values from specification.
const (
	DefaultCheckInterval    = 15 * time.Second
	DefaultTimeout          = 5 * time.Second
	DefaultInitialDelay     = 5 * time.Second
	DefaultFailureThreshold = 1
	DefaultSuccessThreshold = 1

	MinCheckInterval = 1 * time.Second
	MaxCheckInterval = 10 * time.Minute
	MinTimeout       = 100 * time.Millisecond
	MaxTimeout       = 30 * time.Second
	MinInitialDelay  = 0
	MaxInitialDelay  = 5 * time.Minute
	MinThreshold     = 1
	MaxThreshold     = 10
)

// namePattern validates dependency names: lowercase letters, digits, hyphens.
var namePattern = regexp.MustCompile(`^[a-z][a-z0-9-]*$`)

// labelNamePattern validates custom label names per Prometheus naming conventions.
var labelNamePattern = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*$`)

// reservedLabels contains label names that cannot be used as custom labels.
var reservedLabels = map[string]bool{
	"name":       true,
	"group":      true,
	"dependency": true,
	"type":       true,
	"host":       true,
	"port":       true,
	"critical":   true,
}

const (
	minNameLen = 1
	maxNameLen = 63
)

// CheckConfig holds parameters for health check scheduling.
type CheckConfig struct {
	Interval         time.Duration
	Timeout          time.Duration
	InitialDelay     time.Duration
	FailureThreshold int
	SuccessThreshold int
}

// DefaultCheckConfig returns CheckConfig with default values from specification.
func DefaultCheckConfig() CheckConfig {
	return CheckConfig{
		Interval:         DefaultCheckInterval,
		Timeout:          DefaultTimeout,
		InitialDelay:     DefaultInitialDelay,
		FailureThreshold: DefaultFailureThreshold,
		SuccessThreshold: DefaultSuccessThreshold,
	}
}

// Validate checks that config values are within allowed ranges.
func (c CheckConfig) Validate() error {
	if c.Interval < MinCheckInterval || c.Interval > MaxCheckInterval {
		return fmt.Errorf("checkInterval %s out of range [%s, %s]", c.Interval, MinCheckInterval, MaxCheckInterval)
	}
	if c.Timeout < MinTimeout || c.Timeout > MaxTimeout {
		return fmt.Errorf("timeout %s out of range [%s, %s]", c.Timeout, MinTimeout, MaxTimeout)
	}
	if c.Timeout >= c.Interval {
		return fmt.Errorf("timeout %s must be less than checkInterval %s", c.Timeout, c.Interval)
	}
	if c.InitialDelay < MinInitialDelay || c.InitialDelay > MaxInitialDelay {
		return fmt.Errorf("initialDelay %s out of range [%v, %s]", c.InitialDelay, MinInitialDelay, MaxInitialDelay)
	}
	if c.FailureThreshold < MinThreshold || c.FailureThreshold > MaxThreshold {
		return fmt.Errorf("failureThreshold %d out of range [%d, %d]", c.FailureThreshold, MinThreshold, MaxThreshold)
	}
	if c.SuccessThreshold < MinThreshold || c.SuccessThreshold > MaxThreshold {
		return fmt.Errorf("successThreshold %d out of range [%d, %d]", c.SuccessThreshold, MinThreshold, MaxThreshold)
	}
	return nil
}

// Endpoint represents a single network endpoint of a dependency.
type Endpoint struct {
	Host   string
	Port   string
	Labels map[string]string // Custom labels via WithLabel.
}

// Dependency describes a monitored dependency.
type Dependency struct {
	Name      string
	Type      DependencyType
	Critical  *bool // nil = not set (validation error)
	Endpoints []Endpoint
	Config    CheckConfig
}

// ValidateName checks that a dependency name follows the naming rules.
func ValidateName(name string) error {
	if len(name) < minNameLen || len(name) > maxNameLen {
		return fmt.Errorf("invalid dependency name %q: length must be %d-%d", name, minNameLen, maxNameLen)
	}
	if !namePattern.MatchString(name) {
		return fmt.Errorf("invalid dependency name %q: must match [a-z][a-z0-9-]*", name)
	}
	return nil
}

// ValidateLabelName checks that a custom label name is valid.
func ValidateLabelName(name string) error {
	if !labelNamePattern.MatchString(name) {
		return fmt.Errorf("invalid label name: %q", name)
	}
	if reservedLabels[name] {
		return fmt.Errorf("reserved label: %q", name)
	}
	return nil
}

// ValidateLabels checks that all custom labels have valid names.
func ValidateLabels(labels map[string]string) error {
	for k := range labels {
		if err := ValidateLabelName(k); err != nil {
			return err
		}
	}
	return nil
}

// BoolToYesNo converts a bool to "yes"/"no" for the critical label.
func BoolToYesNo(v bool) string {
	if v {
		return "yes"
	}
	return "no"
}

// Validate checks that the dependency configuration is valid.
func (d Dependency) Validate() error {
	if err := ValidateName(d.Name); err != nil {
		return err
	}
	if !ValidTypes[d.Type] {
		return fmt.Errorf("unknown dependency type %q", d.Type)
	}
	if d.Critical == nil {
		return fmt.Errorf("missing critical for dependency %q", d.Name)
	}
	if len(d.Endpoints) == 0 {
		return fmt.Errorf("dependency %q has no endpoints", d.Name)
	}
	for i, ep := range d.Endpoints {
		if ep.Host == "" {
			return fmt.Errorf("missing host for dependency %q endpoint %d", d.Name, i)
		}
		if ep.Port == "" {
			return fmt.Errorf("missing port for dependency %q endpoint %d", d.Name, i)
		}
		if err := ValidateLabels(ep.Labels); err != nil {
			return fmt.Errorf("dependency %q endpoint %d: %w", d.Name, i, err)
		}
	}
	return d.Config.Validate()
}
